### PR TITLE
Use a problem as the intro for newtype pattern

### PR DESCRIPTION
**Pull request by pickfire**
A problem can highlight some of the use cases of newtype
pattern rather than discussing generally that rust is strongly
typed. Make the user understand the need of newtype.

I wonder if it should be in the description instead?

Continued https://github.com/rust-unofficial/patterns/pull/117

r? @simonsan

### CODE REVIEW

File: patterns/newtype.md [hunk 1]
```diff
@@ -97,6 +100,7 @@ Here, `Bar` might be some public, generic type and `T1` and `T2` are some intern
 
 ## See also
 
+- [Advanced Types in the book](https://doc.rust-lang.org/book/ch19-04-advanced-types.html?highlight=newtype#using-the-newtype-pattern-for-type-safety-and-abstraction)
 - [Newtypes in the style guide](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html)
```
Comment (pickfire) [inline patterns/newtype.md:104]: The style guide is outdated or not maintained IIRC.

File: patterns/newtype.md [hunk 2]
```diff
@@ -1,8 +1,11 @@
 # Newtype
 
-Rust has strong static types. This can be very different than what you are used to if you are coming from a loosely-typed language. Don't worry, though. Once you get used to them, you'll find the types actually make your life easier. Why? Because you are making implicit assumptions explicit.
+What if in some cases we want a type to behave similar to another type but
+still different? What if we want to enforce some behavior at compile time
+which using only type aliases is not enough? For example, have a custom
```
Comment (simonsan) [inline patterns/newtype.md:5]:
```suggestion
which using only type aliases is not enough? For example, having a custom
```

File: patterns/newtype.md (hunk 1 above)
Comment (simonsan) [reply patterns/newtype.md:104]: Sad, but true.

```suggestion
```

File: patterns/newtype.md [hunk 3]
```diff
@@ -1,8 +1,11 @@
...
+`Display` implementation for `String` for security reasons (password).
```
Comment (simonsan) [inline patterns/newtype.md:6]:
```suggestion
What if in some cases we want a type to behave similar to another type or 
enforce some behaviour at compile time where using only type aliases would 
not be enough? 

For example, if you want to create a custom `Display` implementation for `String` 
due to security considerations (e.g. passwords).
```

File: patterns/newtype.md [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
 
-A really convenient application of the Rust type system is the Newtype pattern.
+We could use the newtype pattern to provide type safety and encapsulation.
```
Comment (simonsan) [inline patterns/newtype.md:8]:
```suggestion
For aforementioned and such cases we could use the `Newtype` pattern to provide __type safety and encapsulation__.
```

File: patterns/newtype.md (hunk 3 above)
Comment (pickfire) [reply patterns/newtype.md]: The separation looks good but the word "you" does not look that good for me. Either way I just applied it.

File: patterns/newtype.md (hunk 4 above)
Comment (pickfire) [reply patterns/newtype.md:10]:
```suggestion
For such cases we could use the `Newtype` pattern to provide __type safety__ and __encapsulation__.
```

aforementioned is a hard word, I think it is better to remove that.

**Review by simonsan — DISMISSED** [review]

File: patterns/newtype.md [hunk 5]
```diff
@@ -1,8 +1,13 @@
 # Newtype
 
-Rust has strong static types. This can be very different than what you are used to if you are coming from a loosely-typed language. Don't worry, though. Once you get used to them, you'll find the types actually make your life easier. Why? Because you are making implicit assumptions explicit.
+What if in some cases we want a type to behave similar to another type or 
+enforce some behaviour at compile time where using only type aliases would 
+not be enough? 
 
-A really convenient application of the Rust type system is the Newtype pattern.
+For example, if you want to create a custom `Display` implementation for `String`
```
Comment (simonsan) [inline patterns/newtype.md:7]:
```suggestion
For example, if we want to create a custom `Display` implementation for `String` 
```

**Review by simonsan — APPROVED** [review]

**Comment by simonsan** [comment]
Nice, thank you! :-)

**Pull request merged**